datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -0,0 +1,115 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package kafka
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Shopify/sarama"
+)
+
+// TopicHandler is a handler function in order to consume published messages.
+type TopicHandler func(message *sarama.ConsumerMessage) error
+
+// Consumer is a reference structure to subscribe block or trace group produced by EN.
+type Consumer struct {
+	config   *KafkaConfig
+	group    sarama.ConsumerGroup
+	topics   []string
+	handlers map[string]TopicHandler
+}
+
+func NewConsumer(config *KafkaConfig, groupId string) (*Consumer, error) {
+	group, err := sarama.NewConsumerGroup(config.Brokers, groupId, config.SaramaConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &Consumer{
+		config:   config,
+		group:    group,
+		handlers: make(map[string]TopicHandler),
+	}, nil
+}
+
+// Close stops the ConsumerGroup and detaches any running sessions. It is required to call
+// this function before the object passes out of scope, as it will otherwise leak memory.
+func (c *Consumer) Close() error {
+	return c.group.Close()
+}
+
+// AddTopicAndHandler adds a topic associated the given event and its handler function to consume published messages of the topic.
+func (c *Consumer) AddTopicAndHandler(event string, handler TopicHandler) error {
+	if event != EventBlockGroup && event != EventTraceBroup {
+		return fmt.Errorf("the event name must be either 'blockgroup' or 'tracegroup'. given: %v", event)
+	}
+	topic := c.config.getTopicName(event)
+	c.topics = append(c.topics, topic)
+	c.handlers[topic] = handler
+	return nil
+}
+
+// Subscribe subscribes the registered topics with the handlers until the consumer is closed.
+func (c *Consumer) Subscribe(ctx context.Context) error {
+	// TODO-ChainDataFetcher consider error handling if necessary
+	// c.config.SaramaConfig.Consumer.Return.Errors has to be set to true, and
+	// read the errors from c.member.Errors() channel.
+	// Currently, it leaves only error logs as default.
+
+	if len(c.handlers) == 0 || len(c.topics) == 0 {
+		return errors.New("there is no registered handler")
+	}
+
+	// Iterate over consumer sessions.
+	for {
+		if err := c.group.Consume(ctx, c.topics, c); err == sarama.ErrClosedConsumerGroup {
+			return nil
+		} else if err != nil {
+			return err
+		}
+	}
+}
+
+// Setup is run at the beginning of a new session, before ConsumeClaim.
+func (c *Consumer) Setup(s sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+// but before the offsets are committed for the very last time.
+func (c *Consumer) Cleanup(s sarama.ConsumerGroupSession) error {
+	return nil
+}
```

Comment:
Could you explain why those two functions are necessary? It doesn't seem to be used anywhere else.. Is it overloaded elsewhere..?

Reply:
Those functions implements an interface `ConsumerGroupHandler`. In `Subscribe` function, there is a for loop to consume messages and the those functions are triggered in the order of `Setup`, `ConsumeClaims`, and `CleanUp`.

Reply:
It would be better if above comments are written in the function comments, as they do nothing but implements the interface.

Reply:
Added some comments for this. 2684076